src/fsharp/AsyncLazy.fs

```diff
@@ -0,0 +1,134 @@
+namespace Internal.Utilities.Library
+
+open System
+open System.Threading
+open System.Globalization
+
+type private AsyncLazyWeakMessage<'T> =
+    | GetValue of AsyncReplyChannel<Result<'T, Exception>> * CancellationToken
+
+type private AgentInstance<'T> = (MailboxProcessor<AsyncLazyWeakMessage<'T>> * CancellationTokenSource)
+
+[<RequireQualifiedAccess>]
+type private AgentAction<'T> =
+    | GetValue of AgentInstance<'T>
+    | CachedValue of 'T
+
+[<RequireQualifiedAccess>]
+module AsyncLazy =
+
+    // We need to store the culture for the VS thread that is executing now, 
+    // so that when the agent in the async lazy object picks up thread from the thread pool we can set the culture
+    let mutable culture = CultureInfo(CultureInfo.CurrentUICulture.Name)
+
+    let SetPreferredUILang (preferredUiLang: string option) = 
+        match preferredUiLang with
+        | Some s -> 
+            culture <- CultureInfo s
+#if FX_RESHAPED_GLOBALIZATION
+            CultureInfo.CurrentUICulture <- culture
+#else
+            Thread.CurrentThread.CurrentUICulture <- culture
+#endif
+        | None -> ()
+
+[<Sealed>]
+type AsyncLazy<'T> (computation: Async<'T>) =
+
+    let gate = obj ()
+    let mutable computation = computation
+    let mutable requestCount = 0
+    let mutable cachedResult = ValueNone
+    let mutable cachedResultAsync = ValueNone
+
+    let loop (agent: MailboxProcessor<AsyncLazyWeakMessage<'T>>) =
+        async {
+            try
+                while true do
+                    match! agent.Receive() with
+                    | GetValue (replyChannel, ct) ->
+                        Thread.CurrentThread.CurrentUICulture <- AsyncLazy.culture
+                        try
+                            use _reg = 
+                                // When a cancellation has occured, notify the reply channel to let the requester stop waiting for a response.
+                                ct.Register (fun () -> 
+                                    let ex = OperationCanceledException() :> exn
+                                    replyChannel.Reply (Error ex)
+                                )
+
+                            ct.ThrowIfCancellationRequested ()
+
+                            match cachedResult with
+                            | ValueSome result ->
+                                replyChannel.Reply (Ok result)
+                            | _ ->
+                                // This computation can only be canceled if the requestCount reaches zero.
+                                let! result = computation
```

Comment:
For example, this is a bind in an aysnc.  How do I know that this doesn't escape to another thread?  Escape the compilation scope?

As I understand it, you're relying on an unusual invariant that the only places we actually do async thread switches is `PostAndReplyAsync` on the mailbox processors of AsyncLazy nodes?   And so those are the binds that need to be in `asyncErrorLogger { ... }` sections?    And that there are no other async switches anywhere, and all other binds will be synchronous?   To be honest, when I look through the various binds I really have no immediate idea which ones will need to re-establish the computation scope.  It's all quite subtle and risks people writing multi-threaded code inadvertently (a single `Async.Sleep(10)` escapes the compilation scope... Not that we do that but you get the idea)